### ISSUE TITLE
Extend the container instance with custom methods

### ISSUE THREAD

**Issue by guncha** [facebook/relay#1078]
We recently came upon a situation where we needed to add a few extra methods on the container instance to make it conform to a certain interface and it has proven to be more difficult than I anticipated.

To give you a little context, we have a generic `<Form>` component that looks through its children and anything with a `name` prop set is expected to have a `value` accessor that's used to get the component value on form submission. This works for the basic <input> elements as well as any custom ones.

The trouble started when we wanted to use a Relay component in this manner. At first I tried to add the required `value` accessor like this:

``` javascript
class SkillList extends Relay.createContainer(...) {
  get value() {
    return this.refs.component.value
  }
}
```

The reason this doesn't work is that the `Relay.createContainer(...)` call returns a function, the super class in this case, that looks like this:

``` javascript
function ContainerConstructor(props, context) {
  if (!Container) {
    Container = createContainerComponent(Component, spec);
  }
  return new Container(props, context);
}
```

Because of the `return` at the end, the inheritance is reset and `new SkillList()` will actually not return instances of `SkillList`. I'm sure there are good reasons for the lazy class instantiation, I'd just like to learn more about those reasons and see if there's any way to achieve what I'm trying to do cleanly.

While on the topic, I wonder why the API isn't similar to that of Routes and Mutations, i.e.:

``` javascript
class SkillList extends Relay.Container {

  // custom methods here

  static fragments = {
    skills: () => Relay.QL` ... `
  }
}
```

Thank you

**Comment by NevilleS** [comment]
Yeah, it's a bit unfortunate you can't reasonably "decorate" container instances to use them more like regular React components. As it is, they forward props through (which is nice), but that's about the end of the story, otherwise they are a black box...

**Comment by sjchmiela** [comment]
You could use the fact that props are passed down to the child and use callbacks to set parent as a listener to value changes, like this

```Javascript
class SkillList extends Component {
  get value() {
    return this.refs.component.value
  }
  
  render() {
    return (
      <View ref={(ref) => {
        if (ref != null && this.props.setValue) {
          this.props.setValue(ref.value);
        }}
      />
    );
  }
}
```

And then in parent:

```Javascript
class Parent extends Component {
  setValue(value) {
    console.log(`Value of child: ${value}`);
  }

  render() {
    return (
      <SkillList setValue={(value) => this.setValue(value)} />
    );
  }
}
```

**Comment by sjchmiela** [comment]
Or better, pass down function to set ref, and keep it in parent.

```Javascript
class SkillList extends Component {
  get value() {
    return this.refs.component.value
  }
  
  render() {
    return (
      <View ref={(ref) => {
        if (ref != null && this.props.setRef) {
          this.props.setRef(ref);
        }}
      />
    );
  }
}
```

And then in parent:

```Javascript
class Parent extends Component {
  setRef(ref) {
    this._ref = ref;
  }

  render() {
    return (
      <SkillList setRef={(ref) => this.setRef(ref)} />
    );
  }
}
```

**Comment by NevilleS** [comment]
It's certainly possible to build that `Parent` that wraps the `RelayContainer`, but it's awkward. The `RelayContainer` does more that just pass through props, it has methods like `getFragment`, etc. The best possible implementation of a "Wrapper" around the `RelayContainer` would be one that implements the entire `RelayContainer` interface and therefore can be used as the component for a top-level `RelayRenderer`, etc. (see https://github.com/facebook/relay/blob/master/src/container/isRelayContainer.js for how the API is defined right now)

Basically, implementing a wrapper that behaves as if it were a `RelayContainer` seems like a code smell to me - what we really want to do here is extend a `RelayContainer` with some new methods.

**Comment by guncha** [comment]
I'd like to add that it's certainly possible to get this working using controlled components, i.e. the `<Form>` component keeps all of the values in its state and passes them to the input elements, Relay or otherwise, using `value={ ... }`, as well as an `onChange` callback.

The reason I didn't go with this approach is that using uncontrolled components is just simpler - there's no need to merge `onChange` callbacks and the custom form components can manage their own state and not have to merge it with any `value` given to them via props (the component value !== component state for any more sophisticated input).

**Comment by leebyron** [comment]
I'm sorry this issue has sat here for so long. In an attempt to clean up our issue queue we're closing some aging or unclear-action issues. Also, while this issue applies to Relay Classic, our team is trying to focus on Relay Modern going forward.

If this issue is still important to you, please feel free to reopen it with an update.
